### ISSUE TITLE
How to pass amplify app id to the cloud formation template

### ISSUE THREAD

**Issue by vikaskgahlawat** [aws-amplify/amplify-cli#12226]
### Amplify CLI Version

10.8.1

### Question

The lambda function in the backed of my amplify app needs to know the amplify app id to be able to make requests to add subdomains to the app. In the cloudformation template of the lambda function, I do not see any way to pass the amplify app id. There are a few references some of which uses SSM but none is clear. I need to set the amplify app id in the environment variables of my lambda function.

**Comment by ykethan** [comment]
Hey we can add environment variables using the `amplify update function`. Please refer to the [documentation](https://docs.amplify.aws/cli/function/env-vars/) providing this information.
if you are looking to add the environment value dynamically, this is an unsupported workflow.  I will be marking this as feature  request.

As a workaround, you will need to utilize a custom solution by creating parameter store and use hooks to make a API call to the store the parameter to store the value then call the parameter store from the Lambda function to retrieve the value.

**Comment by vikaskgahlawat** [comment]
@ykethan Thanks for your reply. I am fine with adding the environment variables and their values. All I need is the amplify app id that is in the team-provider-info.json to be passed to the cloudformation template. From there, I will set it up as the environment variable of the lambda function. Then, in the CI/CD workflow, there is no team-provider-info.json. How do I get the appId in there.

**Comment by ykethan** [comment]
Hey @vikaskgahlawat, we can use a pre-push hook that inject a file in the `function/src` folder. Then refer the value in the Lambda function code.

for example:
```
const fs = require('fs/promises')
const tpi = require('../team-provider-info.json')
const hookHandler = async (data, error) => {
  const env = data.amplify.environment.envName
  const appid = tpi[env].awscloudformation.AmplifyAppId
  try {
    await fs.writeFile(
      'amplify/backend/function/is1221023b780e7/src/app.json',
      JSON.stringify({ appid }, null, 2),
      'utf-8'
    )
  } catch (cause) {
    throw new Error('Failed to write app.json', { cause })
  }
}
```
or you could create a parameter store then add values in a post push hook using SDK calls and then refer the value in the lambda function using a API call. 
Is the `team-provider-info.json` git ignored?

**Comment by vikaskgahlawat** [comment]
Yes. team-provider-info.json is git ignored. Also, team-provider-info.json does not contain the appId for the amplify apps generated through console or a separate CF template.

**Comment by ykethan** [comment]
is there a reason for adding the team-provider-info.json? is this a open source project? as git ignoring this optional.
you could pull the app down, remove the git ignore, add the hook and push the project. `team-provider-info.json` does not contain sensitive info. https://docs.amplify.aws/cli/reference/files/#team-provider-infojson
anything created in console will need to be pulled for the attributes to appear, you can do this with `amplify pull --<app id> --env <env name>`.

**Comment by vikaskgahlawat** [comment]
Thanks @ykethan for this approach. So even if I remove the team-provider-info.json and use this above script in the pre-push hook, the first deployment of the amplify app will not contain the appID. So it will be a workaround like:
1. Create the amplify app. This will create the apis and functions and other resources.
2. Get the appId from the console manually and do an amplify pull in the developer's machine.
3. Commit the team-provider-info.json and let the amplify CI/CD to update the infrastructure.

I'm not sure if this will really update the function's environment variable but even if it does it's a lengthy workaround or I understood it entirely incorrectly.
But if I understood it correctly, I could instead create an SSM parameter for appId in my app during the very first deployment and after deployment, manually update the SSM parameter value by taking the appId from the console.

Please share your thoughts on this.

**Comment by ykethan** [comment]
hey, if you are creating a new app on the console and attaching a git source, the git source can contain the `team-provider-info.json` and the first create app backend build should run the pre push hook to add (for example `app.json`) file with `appid` to the lambda function src folder. 

for the SSM method, you will need to manually create a parameter store value, then use a post push hook to retrieve the value from `team-provider-info.json` and do a SDK call to the parameter store to store/update the value. in your lambda function you could make a SDK call to get the value and use it.

the above two methods should prevent manually updating the files from console. if the project has already been created and you cannot update the gitignore, this would need to be done on the next push unfortunately.

**Comment by vikaskgahlawat** [comment]
Yes. I can remove my team-provider-info.json from .gitignore. The only drawback I think is that I have to do the deployment twice for every new app creation. I hope AWS provides an easy and straightforward method for this.

**Comment by ykethan** [comment]
hey, I believe there may be some misunderstanding here. can you reach out to me on discord at `ykethan#5706`

**Comment by vikaskgahlawat** [comment]
@ykethan - Yes. I'm pretty sure there is some misunderstanding on my part. Could you please accept my request on discord. It should be from vikasgahlawat#2078

**Comment by ykethan** [comment]
@ykethan i just accepted it

**Comment by vikaskgahlawat** [comment]
One more thought. Team provider info should be categorized first on the basis of appId and then inside it on the basis of env name. It is fairly possible to have two apps having the same env name. 
For example, if two apps are created from the same codebase, then both these apps can contain an env called "dev". How is team-provider-info.json going to handle this.
